brainio/assemblies.py

```diff
@@ -321,4 +359,83 @@ def get_levels(assembly):
         if isinstance(value, IndexVariable) and value.level_names:
             for level in value.level_names:
                 levels.append(level)
-    return levels
\ No newline at end of file
+    return levels
+
+
+class AssemblyLoader:
+    """
+    Loads an assembly from a file.
+    """
+
+    def __init__(self, cls, file_path, group=None, **kwargs):
+        self.assembly_class = cls
+        self.file_path = file_path
+        self.group = group
+
+    def load(self):
+        if self.group is None:
+            data_array = xr.open_dataarray(self.file_path)
+        else:
+            data_array = xr.open_dataarray(self.file_path, group=self.group)
+        result = data_array
+        result = self.assembly_class(data=result)
+        return result
+
+
+class StimulusMergeLoader(AssemblyLoader):
+    """
+    Loads an assembly and add metadata from a stimulus set.
+    """
+
+    def __init__(self, cls, file_path, stimulus_set_identifier=None, stimulus_set=None, **kwargs):
+        super(StimulusMergeLoader, self).__init__(cls, file_path, **kwargs)
+        self.stimulus_set_identifier = stimulus_set_identifier
+        self.stimulus_set = stimulus_set
+
+    def load(self):
+        # data_array = xr.open_dataarray(self.file_path)
+        data_array = super(StimulusMergeLoader, self).load()
+        result = self.merge_stimulus_set_meta(data_array, self.stimulus_set)
+        result = self.assembly_class(data=result)
+        result.attrs["stimulus_set_identifier"] = self.stimulus_set_identifier
+        result.attrs["stimulus_set"] = self.stimulus_set
+        return result
+
+    def merge_stimulus_set_meta(self, assy, stimulus_set):
+        axis_name, index_column = "presentation", "image_id"
+        assy = assy.reset_index(list(assy.indexes))
+        df_of_coords = pd.DataFrame(coords_for_dim(assy, axis_name))
+        cols_to_use = stimulus_set.columns.difference(df_of_coords.columns.difference([index_column]))
+        merged = df_of_coords.merge(stimulus_set[cols_to_use], on=index_column, how="left")
+        for col in stimulus_set.columns:
+            assy[col] = (axis_name, merged[col])
+        return assy
+
+
+class GroupAppendLoader(StimulusMergeLoader):
+
```

Comment:
what's this class for? some comments would be helpful

Reply:
Comments added.